snuba/web/rpc/v1/resolvers/R_eap_spans/common/sampling_in_storage_util.py

```diff
@@ -1,23 +1,140 @@
-) -> HTTPQuerySettings:
-    query_settings = (
-        setup_trace_query_settings() if in_msg.meta.debug else HTTPQuerySettings()
-    )
-    if in_msg.meta.HasField("downsampled_storage_config"):
+
+DOWNSAMPLING_TIER_MULTIPLIERS = {
+    Tier.TIER_512: 1,
+    Tier.TIER_64: 8,
+    Tier.TIER_8: 64,
+    Tier.TIER_1: 512,
+}
+
+
+def _get_time_budget() -> float:
+    sentry_timeout_ms = cast(
+        int, state.get_int_config("sampling_in_storage_sentry_timeout", default=30000)
+    )  # 30s = 30000ms
+    error_budget_ms = cast(
+        int, state.get_int_config("sampling_in_storage_error_budget", default=5000)
+    )  # 5s = 5000ms
+    return sentry_timeout_ms - error_budget_ms
+
+
+def _get_query_duration(timer: Timer) -> float:
+    return timer.get_duration_between_marks("right_before_execute", "execute")
+
+
+def _get_target_tier(timer: Timer) -> Tier:
+    most_downsampled_query_duration_ms = _get_query_duration(timer)
+
+    target_tier = Tier.TIER_512
+    for tier in sorted(Tier, reverse=True)[:-1]:
+        estimated_query_duration_to_this_tier = (
+            most_downsampled_query_duration_ms
+            * cast(int, DOWNSAMPLING_TIER_MULTIPLIERS.get(tier))
+        )
         if (
-            in_msg.meta.downsampled_storage_config.mode
-            == DownsampledStorageConfig.MODE_PREFLIGHT
+            estimated_query_duration_to_this_tier
+            <= _get_time_budget() - most_downsampled_query_duration_ms
         ):
-            query_settings.set_sampling_tier(Tier.TIER_512)
+            target_tier = tier
+    return target_tier
+
+
+def _is_best_effort_mode(in_msg: T) -> bool:
+    return (
+        in_msg.meta.HasField("downsampled_storage_config")
+        and in_msg.meta.downsampled_storage_config.mode
+        == DownsampledStorageConfig.MODE_BEST_EFFORT
+    )
+
+
+def build_snuba_request(
+    request: T, query_settings: QuerySettings, build_query: Callable[[T], Query]
+) -> SnubaRequest:
+    return SnubaRequest(
+        id=uuid.UUID(request.meta.request_id),
+        original_body=MessageToDict(request),
+        query=build_query(request),
+        query_settings=query_settings,
+        attribution_info=AttributionInfo(
+            referrer=request.meta.referrer,
+            team="eap",
+            feature="eap",
+            tenant_ids={
+                "organization_id": request.meta.organization_id,
+                "referrer": request.meta.referrer,
+            },
+            app_id=AppID("eap"),
+            parent_api="eap_span_samples",
+        ),
+    )
+
+
+def run_query_to_correct_tier(
+    in_msg: T,
+    query_settings: HTTPQuerySettings,
+    timer: Timer,
+    build_query: Callable[[T], Query],
+) -> QueryResult:
+    if not in_msg.meta.HasField("downsampled_storage_config"):
+        return run_query(
+            dataset=PluggableDataset(name="eap", all_entities=[]),
+            request=build_snuba_request(in_msg, query_settings, build_query),
+            timer=timer,
+        )
```

Comment:
You should time the estimation query and emit that as a metric so we know how much time this is taking up. You could probably use the timer class for it that is passed in here